### ISSUE TITLE
Rendering out-of-bounds values using OpenGL 1.1 software rendering

### ISSUE THREAD

**Issue by gforney** [firemodels/fds#2294]
```
Please complete the following lines...

FDS Version: 5.5.3
SVN Revision Number: 7031
Compile Date:
Smokeview Version/Revision: 6.1.12
Operating System: Windows

Describe details of the issue below:

The issue is happening while our CFD staff is connected to remote Windows hosts running
Smokeview using Remote Desktop (RDP) clients. This way Smokeview is running with OpenGL
1.1.0 (Software Rendering) that is supported by Windows RDP/RDS protocol.

We noticed the issue since plots/images produced within such a remote Smokeview setup
show unusual coloring of out-of-bounds areas of slices. Data bounds are managed with
Smokeview dialog File/Bounds Slice->Bound data. The reported rendering experience happens
as well while using Smokeview interactively. Given that data exploration is much harder
and any plots with bounds applied are not usable for client documentation purposes.

When running Smokeview on local workstations with OpenGL 3.x/4.X support we do not
experience these changes in coloring while using same FDS jobs. This should emphasize
that same job data is rendered as expected while Smokeview is running on graphics cards
that support native/hardware OpenGL support.

I attached both images, the faulty one renders out-of-bounds area under to the roof
in blue color instead of rendering this area in red.

Please investigate if this behavior could be a bug that only appears if Smokeview runs
on a OpenGL Software Renderer. Any suggestion for mitigation strategies are highly
welcome. 

Attachments:
1. Slice out-of-bounds areas rendered with red color
2. Slice out-of-bounds areas rendered with blue color
3. OpenGL support on local workstation
4. OpenGL support on remote host
5. Smokeview console output on remote indicating OpenGL 1.1 support
6. File/Bounds dialog to show bounding of quantity "temperature"

```

Original issue reported on code.google.com by `sascha.gottfried` on 2015-01-28 20:43:16

<hr>
- _Attachment: slice_rendered_locally.png<br>![slice_rendered_locally.png](https://storage.googleapis.com/google-code-attachments/fds-smv/issue-2336/comment-0/slice_rendered_locally.png)_
- _Attachment: faulty_rendering.png<br>![faulty_rendering.png](https://storage.googleapis.com/google-code-attachments/fds-smv/issue-2336/comment-0/faulty_rendering.png)_
- _Attachment: opengl_support_on_local_workstation.png<br>![opengl_support_on_local_workstation.png](https://storage.googleapis.com/google-code-attachments/fds-smv/issue-2336/comment-0/opengl_support_on_local_workstation.png)_
- _Attachment: opengl_support_on_remote_host.png<br>![opengl_support_on_remote_host.png](https://storage.googleapis.com/google-code-attachments/fds-smv/issue-2336/comment-0/opengl_support_on_remote_host.png)_
- _Attachment: smokeview_output_on_remote_host.png<br>![smokeview_output_on_remote_host.png](https://storage.googleapis.com/google-code-attachments/fds-smv/issue-2336/comment-0/smokeview_output_on_remote_host.png)_
- _Attachment: bound_data_dialog.png<br>![bound_data_dialog.png](https://storage.googleapis.com/google-code-attachments/fds-smv/issue-2336/comment-0/bound_data_dialog.png)_

**Comment by gforney** [comment]
```
(No text was entered with this change)
```

Original issue reported on code.google.com by `mcgratta` on 2015-01-28 20:50:09

**Comment by gforney** [comment]
```
I'll take a look at it
```

Original issue reported on code.google.com by `gforney` on 2015-01-28 21:06:46

**Comment by gforney** [comment]
```
open the "Display Settings" dialog box, then open the "Data coloring" tab.  Click on
Color above max.  In your OpenGL 1.1 version of Smokeview, see if the colors you select
below the "Color above Max" check box are shown as expected on the screen.  If so,
pick the color to match the maximum color ie (255,0,0)
```

Original issue reported on code.google.com by `gforney` on 2015-01-30 21:45:31

**Comment by gforney** [comment]
```
Hi Glen,
thanks for investigating the issue. We already used the approach you suggested. When
I set "color above max" to red, Smokeview renders the "above max" area in blue color.
 Same with Plot3D data. Made another test today with data produced by FDS 6.1.2 - same
experience with OpenGL 1.1.0. 

See attached screenshots.

Glen, keep on suggesting test scenarios that may help you.
```

Original issue reported on code.google.com by `sascha.gottfried` on 2015-02-01 00:03:02

<hr>
- _Attachment: slice_bounding_color_above_max.png<br>![slice_bounding_color_above_max.png](https://storage.googleapis.com/google-code-attachments/fds-smv/issue-2336/comment-4/slice_bounding_color_above_max.png)_
- _Attachment: plot3d_bound_data.PNG<br>![plot3d_bound_data.PNG](https://storage.googleapis.com/google-code-attachments/fds-smv/issue-2336/comment-4/plot3d_bound_data.PNG)_

**Comment by gforney** [comment]
```
Kind reminder - any updates? Do you have any further suggestions for mitigation?
```

Original issue reported on code.google.com by `sascha.gottfried` on 2015-02-06 18:33:10

**Comment by gforney** [comment]
```
what OS is your remote server running?  Windows, Linux or OSX?
```

Original issue reported on code.google.com by `gforney` on 2015-02-06 19:39:57

**Comment by gforney** [comment]
```
Remote OS is Windows Server 2012 R2
```

Original issue reported on code.google.com by `sascha.gottfried` on 2015-02-06 19:56:56

**Comment by gforney** [comment]
```
I take it remote desktop is not an option?  I am able to remote desktop
from home into my windows desktop (not a server) and do not have the OpenGL
problem.  Could you upload the test case that is exhibiting the problem.  I
will try to find a way to duplicate the problem.
```

Original issue reported on code.google.com by `gforney` on 2015-02-06 20:01:10

**Comment by gforney** [comment]
```
Yes, we use "Remote Desktop" (RDP) to connect to remote Windows OS. The remote OS is
running on virtualized hardware (Microsoft Azure - Virtual Machines). Could that make
any difference?

I can not provide the test case immediately since it is a customer case, but I will
discuss with staff how to provide test data for you. BUT: As I already mentioned, I
have noticed the reported behaviour with several FDS cases,that is why I pay closer
attention to the HW/SW setup than the actual FDS case. These properties were always
the same:

* Windows Server 2012 R2 x64 on Microsoft Azure VM 
* Smokeview 6.1.12 x64 running OpenGL 1.1.0
* Windows Remote Desktop clients using RDP 8.X protocol


What do you suggest as a strategy to question my assumption? Thanks in advance!
```

Original issue reported on code.google.com by `sascha.gottfried` on 2015-02-06 20:25:04

**Comment by gforney** [comment]
```
if you have a test case that exhibits the problem that will be fine or if
the customer can simplify the test case (ie remove any proprietary
information) that would be fine also.  When I try to reproduce the case on
my end I want to be sure I am starting with something that results in a
problem.
```

Original issue reported on code.google.com by `gforney` on 2015-02-06 20:27:43

**Comment by gforney** [comment]
```
I made some changes in how I convert data to color when using OpenGL 1.x .  (turned
off data extreme option, contracted colorbar  indices by 2 so that data indices range
from 2->253 instead of 0->255 ).  

Try smokeview revision 21677 at
https://drive.google.com/folderview?id=0B_wB1pJL2bFQc1F4cjJWY2duWTA&usp=sharing

If this doens't fix the problem then the only solution is to use a virtual PC tha t
supports a more modern version of OpenGL

```

Original issue reported on code.google.com by `gforney` on 2015-02-08 17:09:51

**Comment by gforney** [comment]
```
Hi Glenn,
thanks a lot your efforts. I'll give feedback asap.
```

Original issue reported on code.google.com by `sascha.gottfried` on 2015-02-09 11:11:44

**Comment by gforney** [comment]
```
Hi Glenn,
now it works, thank you! But it works on a local test setup that is supposed to be
very similar to the remote cloud setup. I just used "remote desktop" protocol to connect
to a local host running Smokeview (SVN 21677).

fix has to be validated on the production setup. I expect the remote setup to behave
alike. This I will deploy that within this week. 


Attachments:

1. Scene showing slice with bounded data rendered using OpenGL 1.1.0. 
2. Display Settings/Data Coloring dialog shows that "data extreme option" is disabled
if Smokeview is running using OpenGL 1.1.X.


How we apply this change to our systems, what is your suggestion? 

1. We use this version, only for review and testing, in a production workflow?
2. We wait until next official Smokeview release - any date scheduled for 6.1.14?

Best regards
Sascha
```

Original issue reported on code.google.com by `sascha.gottfried` on 2015-02-09 14:24:49

<hr>
- _Attachment: slice_rendered_remote_smokeview_svb_21677.png<br>![slice_rendered_remote_smokeview_svb_21677.png](https://storage.googleapis.com/google-code-attachments/fds-smv/issue-2336/comment-13/slice_rendered_remote_smokeview_svb_21677.png)_
- _Attachment: highlight_extreme_data_disabled.PNG<br>![highlight_extreme_data_disabled.PNG](https://storage.googleapis.com/google-code-attachments/fds-smv/issue-2336/comment-13/highlight_extreme_data_disabled.PNG)_

**Comment by gforney** [comment]
```
I will put out a new "official" version of smokeview within the next week
maybe sooner.

I would like you to do another experiment.  I made two changes in the
smokeview you tested.  I would like to back off one of those changes to try
and identify exactly what failed.  I"ll let you know when I put up another
test code.

The new "official" version is an installer similar to the FDS-Smokeview
installer you used originally.  When you run the smokeview updater it will
copy updated copies of Smokeview and other utilities (smokediff, smokezip
etc) into the directory of your choosing - default is the original
installation directory.  You could do the installation on a local PC and
just copy the bin directory to your virtual PCs.
```

Original issue reported on code.google.com by `gforney` on 2015-02-09 14:32:35

**Comment by gforney** [comment]
```
Hi Glenn,

I am happy you intend to investigate further. I will support your experiments.
```

Original issue reported on code.google.com by `sascha.gottfried` on 2015-02-09 14:47:05

**Comment by gforney** [comment]
```
thanks.  I just posted a smokeview (with offset0 in the name) that turns off the data
extreme option as before but does not "contract" the color bar.  So all color bar indices
between 0 and 255 are present.  
```

Original issue reported on code.google.com by `gforney` on 2015-02-09 16:04:16

**Comment by gforney** [comment]
```
Slice data bounding works perfectly.
```

Original issue reported on code.google.com by `sascha.gottfried` on 2015-02-10 12:09:10

<hr>
- _Attachment: slice_rendered_remote_smokeview_svn_21689.png<br>![slice_rendered_remote_smokeview_svn_21689.png](https://storage.googleapis.com/google-code-attachments/fds-smv/issue-2336/comment-17/slice_rendered_remote_smokeview_svn_21689.png)_

**Comment by gforney** [comment]
```
just to be clear, image in comment 17 was produced with smv_test_21677_win64 on your
virtual azure pcs?

have you tried smv_test_21689_offset0_win64?  
```

Original issue reported on code.google.com by `gforney` on 2015-02-10 12:39:52

**Comment by gforney** [comment]
```
Image in comment #17 was produced with SVN 21689. I added the revision to the image
file names as well. And you can validate smokeview version in top left corner of each
image.
```

Original issue reported on code.google.com by `sascha.gottfried` on 2015-02-10 12:54:13

**Comment by gforney** [comment]
```
I just uploaded an "official" smokeview version 6.1.14 that contains this fix.  Try
it out to make sure that it still works.
thanks
```

Original issue reported on code.google.com by `gforney` on 2015-02-11 21:06:52

**Comment by gforney** [comment]
```
Quick reply: I tried it again on my local test setup using Smokeview 6.1.14. Slice Data
Bounding works as expected. Now we update the cloud setup. I expect same behavior.


Please leave issue open until my next reply. Thanks a lot.
```

Original issue reported on code.google.com by `sascha.gottfried` on 2015-02-12 09:33:20

<hr>
- _Attachment: slice_rendered_remote_smokeview_svn_21718.png<br>![slice_rendered_remote_smokeview_svn_21718.png](https://storage.googleapis.com/google-code-attachments/fds-smv/issue-2336/comment-21/slice_rendered_remote_smokeview_svn_21718.png)_

**Comment by gforney** [comment]
```
Smokeview 6.1.14 does "Slice data bounding" as expected in our actual cloud-based setup.
This was really helpful.

Additional testing revealed that Smokeview 6.0.2 has this issue as well, while 5.6.3
does not have this issue and can even "Highlighting extreme data" using OpenGL 1.1.0.
Let me know if you want to investigate further either in this issue or another. 

Smokeview is awesome.
Best regards from Berlin, Germany.
```

Original issue reported on code.google.com by `sascha.gottfried` on 2015-02-13 17:17:57

<hr>
- _Attachment: slice_rendered_in_cloud_setup_smokeview_svn_21718.png<br>![slice_rendered_in_cloud_setup_smokeview_svn_21718.png](https://storage.googleapis.com/google-code-attachments/fds-smv/issue-2336/comment-22/slice_rendered_in_cloud_setup_smokeview_svn_21718.png)_

**Comment by gforney** [comment]
```
we can investigate further.  I've done two things in 6.1.14, 1) contracted the colorbar
to only use indices 2->253 instead of 0->255 and 2) disable "Highlight extreme data".
 I'll send you some more test smokeviews next week perhaps enabling the highlighting
of extreme data but continuing to use 2->253 .
```

Original issue reported on code.google.com by `gforney` on 2015-02-13 19:18:46

**Comment by gforney** [comment]
```
(No text was entered with this change)
```

Original issue reported on code.google.com by `gforney` on 2015-04-06 17:07:02

**Comment by gforney** [comment]
```
Summary:

"Highlight extreme data" and "Slice Data bounding" features did work in Smokeview 5.6.3
while running on OpenGL 1.1.0

On the road to Smokeview 6.0.2 some code modifications/side effects made the features
behave like reported in this issue (2336). Same applies to all version until 6.1.13

The fix/workaround code originating from this issue is shipped with Smokeview 6.1.14
and fixed the buggy behavior for users in production workflows. But the conclusion
that it is necessary to disable/contract anything may be wrong since these features
were working perfectly in Smokeview 5.6.3

This post from fds-smv discussion group confirms my observation:
https://groups.google.com/forum/?fromgroups#!topic/fds-smv/x5pIbzJAmHg
```

Original issue reported on code.google.com by `sascha.gottfried` on 2015-04-08 09:08:57
